File: Byte&Buy.AngularSPA/src/app/core/dto/cart/cart-item/cart-offer-response.ts
import { Guid } from "guid-typescript";
import { ImageResponse } from "../../image/image-response";
import { MoneyDto } from "../../common/money-dto";

export interface CartOfferResponse{
    id: Guid;
    offerId: Guid;
    image: ImageResponse;
    title: string;
    quantity: number;
    subtotal: MoneyDto;
    type: 'sale' | 'rent';
}

File: Byte&Buy.Backend/src/ByteBuy.Core/DTO/Public/Cart/CartOffer/CartOfferResponse.cs
using ByteBuy.Core.DTO.Public.Image;
using ByteBuy.Core.DTO.Public.Money;
using System.Text.Json.Serialization;

namespace ByteBuy.Core.DTO.Public.Cart.CartOffer;

[JsonPolymorphic(TypeDiscriminatorPropertyName = "type")]
[JsonDerivedType(typeof(SaleCartOfferResponse), "sale")]
[JsonDerivedType(typeof(RentCartOfferResponse), "rent")]
public abstract record CartOfferResponse
{
    public Guid Id { get; init; }
    public Guid OfferId { get; init; }
    public ImageResponse Image { get; init; } = null!;
    public string Title { get; init; } = null!;
    public int Quantity { get; init; }
    public MoneyDto Subtotal { get; init; } = null!;
}

